src/launcher/java/org/truffleruby/launcher/options/CommandLineParser.java

```diff
@@ -268,19 +252,22 @@ private void processArgument() throws CommandLineException {
                     break FOR;
                 case 'J':
                     String javaOption = grabOptionalValue();
-                    jvmArguments.add(argument);
                     final boolean isClasspath = javaOption.equals("-cp") || javaOption.equals("-classpath");
 
                     String javaOptionValue;
                     if (isClasspath) {
                         argumentIndex++;
-                        javaOptionValue = getCurrentArgument();
-                        jvmArguments.add(javaOptionValue);
+                        final String currentArgument = getCurrentArgument();
+                        javaOptionValue = "=" + currentArgument;
                     } else {
                         javaOptionValue = "";
                     }
 
-                    config.getJVMOptions().put(javaOption, javaOptionValue);
+                    arguments.set(argumentIndex - (isClasspath ? 1 : 0), "--jvm." + javaOption.substring(1) + javaOptionValue);
+                    if (isClasspath) {
+                        arguments.remove(argumentIndex);
+                        argumentIndex--;
+                    }
```

Comment:
In this case, I think it would be more readable to duplicate the `arguments.set()` call in each branch as it would simplify the logic at the small cost of duplicating `"--jvm." + javaOption.substring(1)`.